examples/shell/shell_common/cmd_ping_responder.cpp

```diff
@@ -0,0 +1,189 @@
+            ReturnErrorOnFailure(
+                mSessionManager.Init(chip::kTestDeviceNodeId, &chip::DeviceLayer::SystemLayer, &mUDPManager, &mAdmins));
+        }
+        ReturnErrorOnFailure(mExchangeManager.Init(&mSessionManager));
+        ReturnErrorOnFailure(mEchoServer.Init(&mExchangeManager));
+
+        ReturnErrorOnFailure(mSessionManager.NewPairing(peer, chip::kTestControllerNodeId, &mTestPairing,
+                                                        chip::SecureSessionMgr::PairingDirection::kResponder, kAdminId));
+        mEchoServer.SetEchoRequestReceived(EchoServerCmd::HandleEchoRequestReceived);
+        mConnected = true;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Stop()
+    {
+        if (!mConnected)
+        {
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+        if (mTcp)
+        {
+            mTCPManager.Close();
+        }
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+        mUDPManager.Close();
+        mEchoServer.Shutdown();
+        mExchangeManager.Shutdown();
+        mSessionManager.Shutdown();
+        mAdmins.ReleaseAdminId(kAdminId);
+        mAdmins.Reset();
+        mConnected = false;
+
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    static void HandleEchoRequestReceived(chip::Messaging::ExchangeContext * ec, chip::System::PacketBufferHandle payload)
+    {
+        streamer_printf(streamer_get(), "Echo Request, len=%u ... sending response.\n", payload->DataLength());
+    }
+
+    EchoServerCmd(const EchoServerCmd &) = delete;
+    EchoServerCmd & operator=(const EchoServerCmd &) = delete;
+
+    bool mTcp       = false;
+    bool mConnected = false;
+
+    chip::Protocols::Echo::EchoServer mEchoServer;
+    chip::TransportMgr<chip::Transport::UDP> mUDPManager;
+
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    static constexpr size_t kMaxTcpActiveConnectionCount = 4;
+    static constexpr size_t kMaxTcpPendingPackets        = 4;
+    chip::TransportMgr<chip::Transport::TCP<kMaxTcpActiveConnectionCount, kMaxTcpPendingPackets>> mTCPManager;
+#endif // INET_CONFIG_ENABLE_TCP_ENDPOINT
+
+    chip::SecureSessionMgr mSessionManager;
+    chip::SecurePairingUsingTestSecret mTestPairing;
+
+    chip::Messaging::ExchangeManager mExchangeManager;
+
+    chip::Transport::AdminPairingTable mAdmins;
+    chip::Transport::AdminPairingInfo * mAdminInfo     = nullptr;
+    static constexpr chip::Transport::AdminId kAdminId = 0;
+};
+
+EchoServerCmd gEchoServerCmd;
+
+} // namespace
+
+int cmd_ping_responder(int argc, char ** argv)
+{
+    bool tcp         = false;
+    bool disable     = false;
+    bool start       = false;
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (argc == 1 && strcmp(argv[0], "tcp") == 0)
+    {
+        start = true;
+        tcp   = true;
+    }
+    else if (argc == 1 && strcmp(argv[0], "stop") == 0)
+    {
```

Comment:
It seems we can 'ping-responder tcp' or 'ping-responder udp' to start the responder, and we need to explicit issue 'ping-responder stop' to shutdown ping-responder, right?

What if the user forget to issue 'ping-responder stop' after  'ping-responder tcp' or 'ping-responder udp'? I think the user should not be able to use ping or send commands in this case, right?

Reply:
Yes. The consideration is not to block the interactive shell.